include/fiction/algorithms/simulation/sidb/ground_state_space.hpp

```diff
@@ -0,0 +1,896 @@
+                    {
+                        it = this_set.first.erase(it);
+                        this_set.second--;
+                        continue;
+                    }
+
+                    overlap.emplace(*it);
+
+                    ++it;
+                }
+            }
+
+            return overlap.size();
+        }
+    };
+
+    template <sidb_charge_state current_fill_cs>
+    static bool find_valid_witness_partitioning(witness_partitioning_state& st,
+                                                const uint64_t              num_witnesses_for_current_cs) noexcept
+    {
+        if constexpr (current_fill_cs == sidb_charge_state::NEUTRAL)
+        {
+            return st.neutral_witnesses.size() >= num_witnesses_for_current_cs;
+        }
+        else if constexpr (current_fill_cs == sidb_charge_state::POSITIVE)
+        {
+            if (num_witnesses_for_current_cs == 0)
+            {
+                return find_valid_witness_partitioning<sidb_charge_state::NEUTRAL>(st, st.required_neut_count);
+            }
+
+            for (typename witness_partitioning_state::witness_set::const_iterator it = st.positive_witnesses.cbegin();
+                 it != st.positive_witnesses.cend();)
+            {
+                const uint64_t take_witness = *st.positive_witnesses.cbegin();
+
+                st.positive_witnesses.erase(it);
+
+                st.neutral_witnesses.erase(take_witness);
+
+                if (find_valid_witness_partitioning<sidb_charge_state::POSITIVE>(st, num_witnesses_for_current_cs - 1))
+                {
+                    return true;
+                }
+
+                it = ++st.positive_witnesses.emplace(take_witness).first;
+
+                st.neutral_witnesses.emplace(take_witness);
+            }
+
+            return false;
+        }
+        else if constexpr (current_fill_cs == sidb_charge_state::NEGATIVE)
+        {
+            if (num_witnesses_for_current_cs == 0)
+            {
+                return find_valid_witness_partitioning<sidb_charge_state::POSITIVE>(st, st.required_pos_count);
+            }
+
+            for (typename witness_partitioning_state::witness_set::const_iterator it = st.negative_witnesses.cbegin();
+                 it != st.negative_witnesses.cend();)
+            {
+                const uint64_t take_witness = *it;
+
+                st.negative_witnesses.erase(it);
+
+                st.positive_witnesses.erase(take_witness);
+                st.neutral_witnesses.erase(take_witness);
+
+                if (find_valid_witness_partitioning<sidb_charge_state::NEGATIVE>(st, num_witnesses_for_current_cs - 1))
+                {
+                    return true;
+                }
+
+                it = ++st.negative_witnesses.emplace(take_witness).first;
+
+                st.positive_witnesses.emplace(take_witness);
+                st.neutral_witnesses.emplace(take_witness);
+            }
+
+            return false;
+        }
```

Comment:
Possible path with non-void return type!

Reply:
I'm not sure I understand. There are no missing return statements, right?

Reply:
Even if ``current_fill_cs =sidb_charge_state::NONE``, we have a problem, right? 
You can just add for example: 
```c++
else
{
       return false;
}
```

Reply:
True. But this may never happen. This is not publicly accessible code, and the template parameters are just there to give a little edge (if that even does anything). A simple (manual) code flow analysis would show that `sidb_charge_state::NONE` will never occur (in the entire file for that matter). I'm not a fan of useless code branches, and CodeCov isn't either. In any case, I never got a Clang-Tidy/CodeQL warning about this. Only about `sidb_simulation_engine` really.